### ISSUE TITLE
Filter an array of objects by IS NULL or IS NOT NULL does not work properly

### ISSUE THREAD

**Issue by tellezb** [crate/crate#5761]
**CrateDB version**:
1.1.5

**JVM version**:
1.8.0_131

**OS version / environment description**:
Amaon Linux version 4.9.20-11.31.amzn1.x86_64

**Problem description**:
When attempting to filter an array of objects field by IS NULL or IS NOT NULL, improper results are returned. Column definition:
`"product_info" ARRAY(OBJECT (STRICT) AS (
      "amount" DOUBLE,
      "occurrences" INTEGER,
      "product_key" STRING,
      "quantity" DOUBLE
   ))`

In this example, filtering by IS NULL returns values which are both NULL and NOT NULL:
![image](https://user-images.githubusercontent.com/7115725/27342302-a7e12be6-5594-11e7-9c53-df3666ead377.png)

In this example. filter by IS NOT NULL returns nothing:
![image](https://user-images.githubusercontent.com/7115725/27342317-b59cb4c6-5594-11e7-81ef-707b01b220ba.png)

Example column value:
`[{product_key='BD5A79D7-23BA-4B20-96CC-08699A0BE342',occurrences='1',amount='7.38',quantity='1'},{product_key='8AB0DBBD-EAFA-441A-962B-7588706DFA98',occurrences='1',amount='6.76',quantity='1'},{product_key='2157E849-70D0-4587-98CF-7D6F9B9AB114',occurrences='1',amount='6.91',quantity='1'},{product_key='6152ECE4-056F-4906-92C0-7DE1B82AA79F',occurrences='1',amount='7.94',quantity='1'},{product_key='E34633B0-DBAE-4784-ACDD-D65EBF9E5D3D',occurrences='1',amount='4.77',quantity='1'}]`

**Comment by chaudum** [comment]
@tellezb thanks for reporting. we can confirm that there is an issue with IS NULL on object array columns and are working on a fix.

**Comment by turbo-ele** [comment]
Hey @tellezb ,
currently, checking if an object_array is NULL is not supported by crate.
A workaround for this would be to check if a property of the object is null.

This property needs to exist in all objects of the array, otherwise the given row will not be part of the result.

Workaround:
```
SELECT product_info FROM t WHERE product_info['occurrences'] IS NULL;
```

**Comment by tellezb** [comment]
@turbo-ele Thank you for the workaround. Does this mean the bug won't be fixed?

As mentioned, if the property doesn't exist for all objects then the results are still off (using the workaround). This becomes a problem if the object array is dynamic, and each row only uses a subset of all possible objects:

> "payment_info" ARRAY(OBJECT (DYNAMIC) AS (
      "cash" OBJECT (DYNAMIC) AS (
         "amount" DOUBLE,
         "occurrences" LONG
      ),
      "check" OBJECT (DYNAMIC) AS (
         "amount" DOUBLE,
         "occurrences" LONG
      ),
      "credit" OBJECT (DYNAMIC) AS (
         "amount" DOUBLE,
         "occurrences" LONG
      ),
      "debit" OBJECT (DYNAMIC) AS (
         "amount" DOUBLE,
         "occurrences" LONG
      ),
      "giftcard" OBJECT (DYNAMIC) AS (
         "amount" DOUBLE,
         "occurrences" LONG
      ),
[...]
   ))

I haven't been able to come up with a work around for this scenario.

**Comment by turbo-ele** [comment]
@tellezb 
Sorry, that wasn't correct what i said. The properties doesn't need to exist in all the objects, but at least in one object of an array in every row.

Unfortunately, there is no workaround if the array of objects looks completely different on each row. **At least one property needs to be part of any object in your array in any row.**

Here's an example:

```
cr> SELECT * FROM z ORDER BY id;
+----+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| id | product_info                                                                                                                                                                                                             |
+----+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
|  1 | [{"amount": 7.38, "occurrences": 1, "product_key": "BD5A79D7-23BA-4B20-96CC-08699A0BE342", "quantity": 1.0}, {"amount": 6.76, "occurrences": 1, "product_key": "8AB0DBBD-EAFA-441A-962B-7588706DFA98", "quantity": 1.0}] |
|  2 | NULL                                                                                                                                                                                                                     |
|  3 | [{"product_key": "1234"}, {"occurrences": 2}]                                                                                                                                                                            |
|  4 | [{}]                                                                                                                                                                                                                     |
+----+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
SELECT 4 rows in set (0.054 sec)
``` 

If you now query for `product_info['occurrences'] IS NULL` you'll only get rows 2 and 4.
If you query `product_info['occurrences'] IS NOT NULL` the result will be rows 1 and 3 (since both contain an object with property occurrences.

So for the scenario above, you just have to make sure that every object you insert (which is not null) has a certain row applied.



The issue itself is still under investigation.

**Comment by turbo-ele** [comment]
@tellezb the issue will be fixed within the next release.
again, thanks for reporting.

**Comment by turbo-ele** [comment]
@tellezb The issue has been resolved in versions 1.1.6 and 2.0.2
